### DISCUSSION TITLE
Modelleerprincipes - U3. Informatiemodel bruikbaar in object-georienteerde, relationele, en graaf-gebaseerde paradigma's

### DISCUSSION THREAD

**Discussion by pmaria** [Geonovum/disgeo-imsor#55] [General] [answered]
Modelleerprincipe [[U3]](https://geonovum.github.io/disgeo-imsor/modelleerprincipes/#u3-informatiemodel-bruikbaar-in-object-georienteerde-relationele-en-graaf-gebaseerde-paradigma-s) - Het informatiemodel (IMSOR) moet bruikbaar zijn voor het uitdrukken van informatie in in object-georiënteerde, relationele, en graaf-gebaseerde paradigma's

Discussie uit expertgroepsessie:

> - Lennart: En tree gebaseerd?
> - Paul: Bij de modelleermethodes moet ook iets worden gezegd over het belang van validatie mechanismes, en of er mogelijkheden zijn om te valideren voor elke methode/paradigma (want het lijkt belangrijk om te kunnen valideren in het geval van de SOR).
> - Misschien dekt 'graaf' toch niet de lading (GraphQL is bv redelijk OO).
> - Bij het voorbeeld ontstaan er twijfels over het eigenschap 'documentnummer'. Pano ziet het niet als een eigenschap van het pand, Michel merkt op dat het om een gecombineerde externe key lijkt te gaan. Lennart legt uit dat het om een audit trail van een pand gaat, en dat ze destijds de keuze hebben gemaakt om het op deze impliciete wijze te modelleren. Dit had eigenlijk niet in een conceptueel model hoeven staan - maar in een logisch model zou het wel kunnen, aangezien je dan wel met historie/trails kunt werken. De eigenschap/relatie is echter lastig te begrijpen en kan tot silo-denken leiden, dit moeten we zien te vermijden.
> - Pano: de audit trail moet niet vermengt worden met de objectgegevens in hetzelfde informatieobject

**Comment by lvdbrink** [comment, marked as answer]
> Lennart: En tree gebaseerd?

@lennartvanbergen-kadaster bedoelde je hiërarchisch? 

> Paul: Bij de modelleermethodes moet ook iets worden gezegd over het belang van validatie mechanismes, en of er mogelijkheden zijn om te valideren voor elke methode/paradigma (want het lijkt belangrijk om te kunnen valideren in het geval van de SOR).

Of validatie belangrijk is moeten we nog goed overwegen. Gezien de positionering van het "SOR informatiemodel"  - dat een soort semantisch model lijkt te worden dat bovenop de bestaande registraties ligt om samenhang uit te drukken, is het maar de vraag of er dan tegen het IMSOR gevalideerd wordt. De registraties behouden hun eigen aanlevermechanismen en daarbij hun eigen validaties. Validatie is m.i.  aan de afnemerskant minder van belang.

**Reply by lennartvanbergen** [reply]
Ik bedoel: zoals bv. XML en GML. Bv. https://www.w3schools.com/xml/xml_tree.asp ter illustratie dat XML vaak tree-structure genoemd wordt.

**Reply by pmaria** [reply]
Ik denk dat hierarchisch dan een generiekere term is. Alleen weet ik niet of we het als paradigma moeten onderscheiden. 

XML is maar een manier om informatie hierarchisch te serialiseren. JSON is een andere. Maar er zijn hierarchische serialisaties voor het graafgebaseerde paradigma. Ook in XML en JSON bijv.

**Reply by lennartvanbergen** [reply]
Prima, zeker als je erbij zet: "bv. JSON en XML", dan snapt iedereen het.

**Comment by lvdbrink** [comment]
Over dit uitgangspunt zijn we het na de toevoeging van het hierarchische paradigma wel eens. We moeten alleen nog even checken bij Paul (die bij de expertgroepbijeenkomst van 17/6 afwezig was) wat hij met zijn opmerking over validatie bedoelt in deze context.

**Comment by pmaria** [comment]
@PalmJanssen tijdens de discussie over dit uitgangspunt had jij het volgende opgemerkt:

> Paul: Bij de modelleermethodes moet ook iets worden gezegd over het belang van validatie mechanismes, en of er mogelijkheden zijn om te valideren voor elke methode/paradigma (want het lijkt belangrijk om te kunnen valideren in het geval van de SOR).

Weet je nog wat je hiermee bedoelde? Zouden we dit moeten verwerken in dit uitgangspunt, of elders in het document? Zo ja, hoe zou je dat terug willen zien?

**Reply by PalmJanssen** [reply]
Je kan modellen zo maken dat het valideren of data conform het model zijn complex is of maar beperkt mogelijk. Hoe flexibeler het model hoe moeilijker de validatie is. Je kan bijvoorbeeld een model maken dat als een soort metaconstructie is gemaakt voor klasse - attribuut - relatie binding waarbij attributen in waardelijsten zijn opgenomen. Een dergelijk model geeft veel ruimte en flexibiliteit maar geeft geen strakke validatieregels. Als validatie een uitganspunt is dan heeft dit effect op de wijze van modelleren.

**Reply by pmaria** [reply]
Ok, ik weet alleen niet zo goed hoe we dat zouden moeten verwerken in de modelleerprincipes. In principe gaan we uit van MIM. Dat beperkt volgens mij al een significant deel van de modelleerconstructies. Missen we dan nog steeds een uitgangspunt / principe rondom valideerbaarheid?

**Reply by PalmJanssen** [reply]
MIM heeft daar niet zoveel mee te maken. Je kan modellen heel strak modelleren bijvoorbeeld met waardeliijsten en constraints. Of meer open met vrij in te vullen velden en alles optioneel. Daarnaast (maar dan is ook het eene implementatieformat geschikter voor validatie dan het andere. Dan hebben we het natuurlijk niet meer over modelleren maar implementeren.

**Comment by lvdbrink** [comment]
Ik heb even gekeken of validatie in de [Architectuurbeschrijving Voorzieningen Samenhangende Objectenregistratie](https://docs.geostandaarden.nl/disgeo/arch/) beschreven wordt. Daar wordt het niet uitgebreid beschreven maar wel genoemd als onderdeel van registratie: 

> Registreren = Het valideren en vastleggen van objectgegevens in de registratie.
> [...]
> de validatie die tijdens registratie wordt gedaan is ook als zelfstandige functionaliteit bruikbaar om objectgegevens te kunnen valideren zonder ze te registreren.
> [...]
> Bij elke toevoeging of wijziging van een gegeven vindt vooraf validatie aan de gegevensregels plaats. Alleen valide gegevens worden geregistreerd.

In de [eisen aan de inhoud](https://docs.geostandaarden.nl/disgeo/emso/) wordt het alleen genoemd in de context van informatieproducten.

In de architectuurbeschrijving wordt validatie kortom genoemd als onderdeel van registreren, dus het informatiemodel moet dit wel ondersteunen of er moeten validatieregels uit afgeleid of aan toegevoegd kunnen worden. Ik kan me voorstellen dat de validatieregels niet in elk modelleerparadigma IN het informatiemodel zitten maar ook een apart product kunnen zijn. Hoe zou je dit kunnen verwoorden?

**Comment by lennartvanbergen** [comment]
Paradigma's of moet uit te drukken zijn in technische formaten die ...

**Comment by lennartvanbergen** [comment]
Het voorbeeld: document uit BAG roept bij mij discussie op, of het voorbeeld wel treffend is t.a.v. paradigma. 

Het betreft een sleutel/relatie naar een ander object, Document. Document zit niet in de BAG, maar de relatie ernaartoe wel. 

Het voorbeeld kan vanuit het perspectief: document is geen eigenschap van Pand. Maar dan hoort er wel bij: hoe modelleer je dan een relatie naar een ander objecttype? 

Helaas is het zo dat de relatie naar document niet alleen de id is maar ook de datum. Dit kan je in OO  maar ook in andere paradigma's lastig modelleren, want het is een relatie, met een datum. Omdat dit ook niet goed in een ander paradigma is te modelleren, is het voorbeeld niet helemaal fijn. 

Eens dat een relatie naar Document niet in Pand hoort, als eigenschap, omdat het geen eigenschap van Pand is. Het hoort in een logisch model, als audittrail - metadata net zoals historie. De verantwoording welk besluit heeft geleid tot aanpassing van de basisregistratie. 

Nou zijn relaties ook eigenschappen. Vanuit deze optiek kan je stellen: de relatie naar document is geen eigenschap van pand. Althans niet op CIM niveau. Maar op Logisch niveau hoort deze relatie er wel bij. Het is een essentieel gegeven van de BAG, de verwijzing naar het besluit/document. 

Hoe modelleer je dit dan op logisch niveau, als SOR logisch is? 

Antwoord: als metadata. Want het is geen eigenschap van Pand. 

Deze rationale mis ik alleen in het voorbeeld.

**Comment by lennartvanbergen** [comment]
Document is in de BAG geen object (zie BAG wet), het is een relatie naar een Document, en het Document zit buiten de BAG (bv. via MIM: externe koppeling). Dit valt dus niet te modelleren als Document. Het is een relatie naar een Document. 

Op logisch niveau mag je dingen toevoegen bij Pand als dat handig is voor de implementatie. Op conceptueel IM niveau hoort document mijn inziens inderdaad niet bij Pand. Maar logisch mag het gewoon. 

Ik denk dat we toe moeten willen naar: 
- op conceptueel IM niveau zijn alle eigenschappen directe eigenschappen 
- omdat het mag op logisch niveau, mag je document toevoegen bij Pand, en dan wil je denk ik metadata kunnen toevoegen op logisch IM niveau om aan te duiden of het een directe eigenschap is.
